### DISCUSSION TITLE
Submenu bound to collection- How to remove the gap between menu items

### DISCUSSION THREAD

**Discussion by Sciolo** [AvaloniaUI/Avalonia#12021] [Q&A] [answered]
I have a flyout menu that contains a submenu that is bound to a collection. It's all working well, but there is a gap in between items on the submenu that I can't get rid of. The issue is that when users click in the gap nothing happens other than the menu closing, which makes the application seem dodgy. If someone could help me work out how to remove the unwanted gap then I'd really appreciate it thanks.

I'm using Avalonia 11 RC2.2 on Windows.

I've set the background of the buttons on the submenu to blue to make it easy to see the gaps in the attached image.

Kind Regards,
James

![submenu-bound-to-itemssource](https://github.com/AvaloniaUI/Avalonia/assets/53260221/ec6227aa-2907-4e21-99c3-3db1188f680f)

![button-flyout-source-code](https://github.com/AvaloniaUI/Avalonia/assets/53260221/87a61fa5-ee8c-4eb4-9e62-dfe033b5c894)

**Comment by Sciolo** [comment]
I've done the following as a workaround. Almost entirely removed the gap using negative margins. Not a decent solution, but I could roll it out like this (with the original image and minus the blue background on the MenuItem).

![presets-flyout-menu](https://github.com/AvaloniaUI/Avalonia/assets/53260221/16bfb4cf-0537-4aa2-8393-11c7bc418910)

![presets-flyout-menu-source-code](https://github.com/AvaloniaUI/Avalonia/assets/53260221/a01949b3-2f21-4047-956e-e13f1a86d821)

Regards,
James

**Reply by maxkatz6** [reply]
If you need to adjust styles of default controls, easier to write your simple styles with selectors for that.
Please use DevTools to see what properties you can adjust and write corresponding styles.

https://docs.avaloniaui.net/docs/styling/styles
https://docs.avaloniaui.net/docs/getting-started/developer-tools

**Reply by Sciolo** [reply]
I usually do create styles with selectors, but haven't been able to workout where that darn margin or padding is coming from. I will try again.

Thanks,
James

**Reply by jp2masa** [reply, marked as answer]
You shouldn't create a `MenuItem` in the `DataTemplate`, as you will get a `MenuItem` inside another `MenuItem`. Try using `ItemContainerTheme` to set the properties instead (e.g. https://github.com/AvaloniaUI/Avalonia/discussions/10326#discussioncomment-4967899).

**Comment by Sciolo** [comment]
Thanks jp2masa, There are no issues with unwanted margins/padding when using a ItemContainerTheme. Some source code follows in case it is useful to someone else. I haven't got the Icon image working yet, but that is another issue.

```
            <Button Grid.Column="1" Grid.Row="0" Classes="presetButton" Padding="1,0,0,0"
                    IsEnabled="{Binding Path=!#AspectsDataGridControl.IsLocked}">
                <Image Source="/Assets/Images/preset-menu-button-image.png" MinWidth="7" MinHeight="12" Height="12"
                       HorizontalAlignment="Center" VerticalAlignment="Center" />
                <Button.Flyout>
                    <MenuFlyout Placement="RightEdgeAlignedTop">
                        <MenuItem Header="New Preset" Command="{Binding Path=#AspectsDataGridControl.NewPresetCmd}" />
                        <MenuItem Header="Clear Default"
                                  Command="{Binding Path=#AspectsDataGridControl.ClearDefaultPresetCmd}" />
                        <MenuItem Header="Presets"
                                  ItemsSource="{Binding Path=#AspectsDataGridControl.PresetItemsSource, Mode=OneWay}">
                            <MenuItem.ItemContainerTheme>
                                <ControlTheme TargetType="MenuItem" x:DataType="serviceInterfaces:IPreset"
                                              BasedOn="{StaticResource {x:Type MenuItem}}">
                                    <Setter Property="Padding" Value="15,5,0,5" />
                                    <Setter Property="Header" Value="{Binding Path=Name, Mode=OneWay}" />
                                    <Setter Property="Command"
                                            Value="{Binding Path= #AspectsDataGridControl.ApplyPresetCmd}" />
                                    <Setter Property="CommandParameter" Value="{Binding Path=., Mode=OneWay}" />
                                    <Setter Property="ContextMenu">
                                        <Setter.Value>
                                            <ContextMenu>
                                                <MenuItem Header="Set as Default"
                                                          Command="{Binding Path= #AspectsDataGridControl.SetDefaultPresetCmd}"
                                                          CommandParameter="{Binding Path=., Mode=OneWay}" />
                                                <MenuItem Header="Rename Preset"
                                                          Command="{Binding Path= #AspectsDataGridControl.RenamePresetCmd}"
                                                          CommandParameter="{Binding Path=., Mode=OneWay}" />
                                                <MenuItem Header="Delete Preset"
                                                          Command="{Binding Path= #AspectsDataGridControl.DeletePresetCmd}"
                                                          CommandParameter="{Binding Path=., Mode=OneWay}" />
                                            </ContextMenu>
                                        </Setter.Value>
                                    </Setter>
                                </ControlTheme>
                            </MenuItem.ItemContainerTheme>
                        </MenuItem>
                    </MenuFlyout>
                </Button.Flyout>
            </Button>
```
